File: problems/01-10-2025-S1-Trees/226-invert-binary-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */


 /**
  * Solution DFS recursive:
  * 
  * Traverse the subtrees until you can't and then perform swaps
  * 
  */

class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        // Base case, inverse of an empty tree is an empty tree
        if (root == nullptr) return nullptr;
        // recursively traverse each subtree
        TreeNode* left = invertTree(root->left);
        TreeNode* right = invertTree(root->right);
        // Once we cant traverse left or right we perform a swap
        root->left = right;
        root->right = left;
        // Once the swap is performed we backtrack
        return root;
    }
};

File: problems/01-10-2025-S1-Trees/226-invert-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        
        # edge case: when the current node is None 
        if not root:
            return

        # swap the left and right child of the current node
        root.left, root.right = root.right, root.left
        # recursively call on left and right child
        self.invertTree(root.left)
        self.invertTree(root.right)
        
        return root
    
    